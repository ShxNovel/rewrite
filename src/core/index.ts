export * from './chapter';

export * from './motion';
export * from './text';
