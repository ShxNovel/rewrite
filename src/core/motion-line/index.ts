export * from "./motion-line";
export * from "./motion-addon";
export * from "./item";
