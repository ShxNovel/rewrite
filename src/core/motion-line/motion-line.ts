import { RewriteItem } from "./item";
import {
    MotionAddon,
    UserMotionMethods,
    UserMotionUnitType,
} from "./motion-addon";

export interface MotionUnitType extends UserMotionUnitType {}

export type MotionUnit = {
    type: keyof MotionUnitType;
    args: Record<PropertyKey, unknown>;
};

export type MotionMethodLike = (...args: unknown[]) => MotionLine;

export interface MotionMethods {
    r(name: string, scene?: string): MotionLine;
    r(item: RewriteItem): MotionLine;
}

export interface MotionLine extends MotionMethods, Partial<UserMotionMethods> {
    motionLine: MotionUnit[];
}

const MotionMethodsImpl = {
    /** side effect */
    r(this: MotionLine, arg1: RewriteItem | string, scene: string = "main") {
        if (this.motionLine.length === 0) {
            throw new Error(
                "MotionLine.r() called with no previous motion unit"
            );
        }

        if (typeof arg1 === "object") {
            scene = arg1.scene ?? "main";
            arg1 = arg1.name;
        }

        const unit = this.motionLine[this.motionLine.length - 1];
        unit.args.name = arg1;
        unit.args.scene = scene;

        return this;
    },
};

export function rMotionLine() {
    const result: MotionLine = {
        motionLine: [] as MotionUnit[],

        ...MotionMethodsImpl,
        ...MotionAddon,
    };

    return result;
}
