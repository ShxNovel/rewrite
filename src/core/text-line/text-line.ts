import { TextAddon, UserTextMethods, UserTextUnitType } from "./text-addon";

export type TextUnitType = "Text" | "Pause" | UserTextUnitType;

export type TextUnit = {
    type: TextUnitType;
    args: Record<PropertyKey, unknown>;
};

export type TextMethodLike = (...args: unknown[]) => TextLine;

export interface TextMethods {
    text(word: string): TextLine;
    pause(ms: number): TextLine;
}

export interface TextLine extends TextMethods, Partial<UserTextMethods> {
    textLine: TextUnit[];
}

const TextMethodsImpl = {
    text(this: TextLine, word: string) {
        this.textLine.push({
            type: "Text",
            args: { word },
        });
        return this;
    },
    pause(this: TextLine, ms: number) {
        this.textLine.push({
            type: "Pause",
            args: { ms },
        });
        return this;
    },
};

export function rTextLine() {
    const result: TextLine = {
        textLine: [] as TextUnit[],

        ...TextMethodsImpl,
        ...TextAddon,
    };

    return result;
}
