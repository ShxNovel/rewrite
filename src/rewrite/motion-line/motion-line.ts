import { RewriteItem } from "./item";
import {
    MotionAddon,
    UserMotionMethods,
    UserMotionUnitType,
} from "./motion-addon";

export type MotionUnitType = "Text" | "Pause" | UserMotionUnitType;

export type MotionUnit = {
    type: MotionUnitType;
    args: Record<PropertyKey, unknown>;
};

export type MotionMethodLike = (...args: unknown[]) => MotionLine;

export interface MotionMethods {
    r(name: string, scene?: string): MotionLine;
    r(item: RewriteItem): MotionLine;
}

export interface MotionLine extends MotionMethods, Partial<UserMotionMethods> {
    motionLine: MotionUnit[];
}

export function rMotionLine() {
    const result: MotionLine = {
        motionLine: [] as MotionUnit[],

        r(arg1: RewriteItem | string, scene: string = "main") {
            if (this.motionLine.length === 0) {
                throw new Error(
                    "MotionLine.r() called with no previous motion unit"
                );
            }

            if (typeof arg1 === "object") {
                scene = arg1.scene ?? "main";
                arg1 = arg1.name;
            }

            const unit = this.motionLine[this.motionLine.length - 1];
            unit.args.name = arg1;
            unit.args.scene = scene;

            return this;
        },

        ...MotionAddon,
    };

    return result;
}
