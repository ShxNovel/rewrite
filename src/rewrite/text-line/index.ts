export * from "./text-line";
export * from "./text-addon";
