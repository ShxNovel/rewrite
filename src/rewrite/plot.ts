import { makeArgs, mergeArgs } from "./tools";
import { rewritePush } from "./rewrite";

import type { TextLine } from "./text-line";
import { rTextLine } from "./text-line";
import { rMotionLine, type MotionLine } from "./motion-line";

export type RewriteInfo = {
    title?: string;
    icon?: string;
};

/**
 * Creates a plot information object;
 * It can be used by {@link RewritePlot.use}.
 */
export function rInfo(userArgs: Partial<RewriteInfo>): Partial<RewriteInfo> {
    return makeArgs({}, userArgs);
}

export type RewritePlotAttribute = {
    /** The type of plot, defaults to "Talk" */
    type: "Talk" | "Aside" | "Choice";

    label?: string;
} & RewriteInfo;

export type RewritePlotLine = {
    text: TextLine;
    motion: MotionLine;
};

export type RewritePlot = RewritePlotAttribute &
    RewritePlotLine & {
        /**
         * Initializes a new plot instance
         * and pushes the plot to rewrite stack
         */
        make(args: Partial<RewritePlotLine>): RewritePlot;

        /**
         * With side-effects. \
         * Apply user arguments to current plot configuration
         */
        use(args: Partial<RewritePlotAttribute>): RewritePlot;
    };

export function rPlot(userArgs: Partial<RewritePlot>) {
    const defaultPlot: RewritePlot = {
        type: "Talk",

        text: rTextLine(),
        motion: rMotionLine(),

        make(args: Partial<RewritePlotLine> = {}) {
            const result = mergeArgs(this, args);
            rewritePush(result);
            return result;
        },

        use(args) {
            return Object.assign(this, args);
        },
    };

    return mergeArgs(defaultPlot, userArgs);
}
