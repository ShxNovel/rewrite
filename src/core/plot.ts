import { makeArgs, mergeArgs } from "./tools";
import { rewritePush } from "./rewrite";

import type { TextLine } from "./text-line";
import { rTextLine } from "./text-line";
import { rMotionLine, type MotionLine } from "./motion-line";

export type Info = {
    title?: string;
    icon?: string;
};

/**
 * Creates a plot information object;
 * It can be used by {@link Plot.use}.
 */
export function rInfo(userArgs: Partial<Info>): Partial<Info> {
    return makeArgs({}, userArgs);
}

export type PlotAttribute = {
    /** The type of plot, defaults to "Talk" */
    type: "Talk" | "Aside" | "Choice";

    label?: string;
} & Info;

export type PlotLine = {
    text: TextLine;
    motion: MotionLine;
};

export type BasicPlot = PlotAttribute & PlotLine;
export type Plot = BasicPlot & PlotEffect & PlotPure;

export type PlotPure = {
    /** Creates a new `Talk` plot instance. */
    get talk(): Plot;
    /** Creates a new `Aside` plot instance. */
    get aside(): Plot;
    /** Creates a new `Choice` plot instance. */
    get choice(): Plot;
};

export type PlotEffect = {
    /**
     * Initializes a new plot instance
     * and pushes the plot to rewrite stack
     */
    begin(args?: Partial<BasicPlot>): Plot;
    use(args: Partial<BasicPlot>): Plot;
    useText(args: TextLine): Plot;
    useMotion(args: MotionLine): Plot;
    useInfo(args: Partial<PlotAttribute>): Plot;
};

const PlotEffect: PlotEffect = {
    begin(this: Plot, args = {}) {
        const base = rPlot();
        const result = mergeArgs(mergeArgs(base, this), args);
        rewritePush(result);
        return result;
    },
    use(this: Plot, args) {
        return Object.assign(this, args);
    },

    useText(this: Plot, args) {
        const object: Partial<PlotLine> = { text: args };
        Object.assign(this, object);
        return this;
    },

    useMotion(this: Plot, args) {
        const object: Partial<PlotLine> = { motion: args };
        Object.assign(this, object);
        return this;
    },

    useInfo(this: Plot, args) {
        return Object.assign(this, args);
    },
};

export function rPlot(userArgs: Partial<Plot> = {}): Plot {
    const ResultPure: PlotPure = Object.defineProperties({} as PlotPure, {
        talk: {
            get() {
                const base = rPlot();
                const temp: Plot = mergeArgs(base, this);
                return temp.useInfo({ type: "Talk" });
            },
            enumerable: false,
        },
        aside: {
            get() {
                const base = rPlot();
                const temp: Plot = mergeArgs(base, this);
                return temp.useInfo({ type: "Aside" });
            },
            enumerable: false,
        },
        choice: {
            get() {
                const base = rPlot();
                const temp: Plot = mergeArgs(base, this);
                return temp.useInfo({ type: "Choice" });
            },
            enumerable: false,
        },
    });

    const result = Object.assign(ResultPure, {
        type: "Talk" as PlotAttribute["type"],

        text: rTextLine(),
        motion: rMotionLine(),

        ...PlotEffect,
    }) satisfies Plot;

    return mergeArgs(result, userArgs);
}
