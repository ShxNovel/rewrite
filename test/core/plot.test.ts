import { assert, expect, test } from "vitest";
import { rInfo, rPlot } from "../../src/core/plot";
import { rTextLine } from "../../src/core/text-line";
import { rMotionLine } from "../../src/core/motion-line";

test("rInfo", () => {
    const i1 = rInfo({ icon: "name", title: "title" });
    const i2 = rInfo({ icon: "name", title: "title" });

    expect(i1).toEqual(i2);
});

test("rPlot", () => {
    const p1 = rPlot({ icon: "name", title: "title" });
    const p2 = rPlot({ icon: "name", title: "title" });

    expect(p1).toEqual(p2);
});

test("rPlot.useInfo", () => {
    const p1 = rPlot({ icon: "name", title: "title" });
    const p2 = rPlot({ icon: "name", title: "title" });

    const p1use = p1.useInfo({ label: "label", icon: "icon" });
    const p2use = p2.useInfo({ label: "label", icon: "icon" });

    const i1 = rInfo({ icon: "name", title: "title" });

    expect(p1use).toEqual(p2use);
    expect(p1use).toEqual(p1use.useInfo({ label: "label", icon: "icon" }));
    expect(p1use).toEqual(p1use.useInfo(i1));
});

test("rPlot.begin", () => {
    const p1 = rPlot({ icon: "name", title: "title" });
    const p2 = rPlot({ icon: "name", title: "title" });

    const p1amke = p1.begin({
        text: rTextLine(),
        motion: rMotionLine(),
    });
    const p2amke = p2.begin({
        text: rTextLine(),
        motion: rMotionLine(),
    });

    const t1 = p1amke.begin({
        text: rTextLine().text("11").pause(1000),
        motion: rMotionLine(),
    });

    const t2 = p2amke.begin({
        text: rTextLine().text("11").pause(1000),
        motion: rMotionLine(),
    });

    expect(p1amke).toEqual(p2amke);
    expect(t1).toEqual(t2);
});

test("rPlot.type", () => {
    const p1 = rPlot({ icon: "name", title: "title" });
    const p11 = p1.talk;
    const p21 = p11.aside;
    const p31 = p11.choice;

    // no side effect
    expect(p1).toStrictEqual(p11);
    expect(Object.is(p1, p11)).toEqual(false);

    // other check
    expect(p1).toMatchSnapshot();
    expect(p11).toMatchSnapshot();
    expect(p21).toMatchSnapshot();
    expect(p31).toMatchSnapshot();
});
