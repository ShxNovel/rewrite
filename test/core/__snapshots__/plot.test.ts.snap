// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`rPlot.type 1`] = `
{
  "begin": [Function],
  "icon": "name",
  "motion": {
    "motionLine": [],
    "r": [Function],
  },
  "text": {
    "pause": [Function],
    "text": [Function],
    "textLine": [],
  },
  "title": "title",
  "type": "Talk",
  "use": [Function],
  "useInfo": [Function],
  "useMotion": [Function],
  "useText": [Function],
}
`;

exports[`rPlot.type 2`] = `
{
  "begin": [Function],
  "icon": "name",
  "motion": {
    "motionLine": [],
    "r": [Function],
  },
  "text": {
    "pause": [Function],
    "text": [Function],
    "textLine": [],
  },
  "title": "title",
  "type": "Talk",
  "use": [Function],
  "useInfo": [Function],
  "useMotion": [Function],
  "useText": [Function],
}
`;

exports[`rPlot.type 3`] = `
{
  "begin": [Function],
  "motion": {
    "motionLine": [],
    "r": [Function],
  },
  "text": {
    "pause": [Function],
    "text": [Function],
    "textLine": [],
  },
  "type": "Aside",
  "use": [Function],
  "useInfo": [Function],
  "useMotion": [Function],
  "useText": [Function],
}
`;

exports[`rPlot.type 4`] = `
{
  "begin": [Function],
  "motion": {
    "motionLine": [],
    "r": [Function],
  },
  "text": {
    "pause": [Function],
    "text": [Function],
    "textLine": [],
  },
  "type": "Choice",
  "use": [Function],
  "useInfo": [Function],
  "useMotion": [Function],
  "useText": [Function],
}
`;
